### ISSUE TITLE
MTNG 3Gpc/h inference

### ISSUE THREAD

**Issue by maho3** [maho3/ltu-gobig#25]

**Comment by maho3** [comment]
The reason for this ^ was that the MTNG simulation data was already masked in a conical lightcone. Specifically, it covered one octant of the sky.
![image](https://github.com/user-attachments/assets/baca709f-374e-4421-80e9-728e1d620c84)

When we measured P(k) without the mask correction, this messed with the large-scale modes of the galaxy P(k), and removed the classic 'turnover' we expect of large-scale surveys.

Instead, I have taken this raw MTNG lightcone and rotated it to overlap with the NGC footprint.
![image](https://github.com/user-attachments/assets/5c86f896-6fe7-4888-9ec6-406678adf793)

And then trimmed things down in the observational sample to only include the overlap.
![image](https://github.com/user-attachments/assets/e25ecc0b-6a44-4dab-b421-94a68f312cf2)

After redshift reweighting, this matches well:
![image](https://github.com/user-attachments/assets/d5fdd8a6-dcbd-46af-bf0a-0bd8a076da71)

Now, the monopole P(k) looks closer to the observational sample
![image](https://github.com/user-attachments/assets/a5afa22c-ca6f-4330-8417-b2abf196be04)

Though, the MTNG cosmology + galaxy configuration here may not be ideal.

**Comment by suicee** [comment]
## Coverage  
Monopole Pk of CMASS simulations (black) and MTNG (red), with $k_{\text{max}} \approx 0.3 $.
![image](https://github.com/user-attachments/assets/79573a4e-26b4-432c-9566-1e3b6d4d0f6d)

## Inference 
Posterior from an NPE model trained on CMASS pk and applied to MTNG data.
![image](https://github.com/user-attachments/assets/42582a5a-cc6f-4731-a8d7-a86f797ee970)

**Comment by amariad** [comment]
Results from a quick Random Forest regression trained on the P(k) (with kmax~0.4) measured from the simulated cmass lightcones, to estimate the cosmo parameters and compare to MTNG (orange star).

[RF_ltu_mtnglike_fastpm_3gpc_param_estimation.pdf](https://github.com/user-attachments/files/17066035/RF_ltu_mtnglike_fastpm_3gpc_param_estimation.pdf)

**Comment by amariad** [comment]
![image](https://github.com/user-attachments/assets/5fbb2a30-be5e-49e1-abbe-0dcfe07aa9f4)
Galaxy power spectra run from 500Mpc/h sub volumes  of quijotelike runs with varying HOD models, compared to that of the massless neutrino, MTNG 500Mpc/h volume.

**Comment by amariad** [comment]
![image](https://github.com/user-attachments/assets/b8c9c6bc-e728-4524-94e7-77b6670be4bd)
<img width="256" alt="image" src="https://github.com/user-attachments/assets/c17ef1a9-43fe-4cee-8229-bd505327151c">


A selection of Pk which have values of cosmo params most similar to MTNG across all 5 cosmo params. Here we can see how varying the HOD is really affecting the Pk.

**Comment by amariad** [comment]
![image](https://github.com/user-attachments/assets/148ebf71-ef7c-4bb3-acdc-acf49002ce87)

RF estimation trained on updated quijotelike subvolume hod Pk. Is now more accurately predicting MTNG. note: this is for cosmo pram estimation and each realization has 10 hods with the same cosmo params

**Comment by maho3** [comment]
3 Gpc/h lightcone inference

Previously no constraints on HOD params:
![image](https://github.com/user-attachments/assets/6b47fd9b-f7b5-4f99-9a42-b45153ec9cfe)

Now with HOD bug fixed:
![image](https://github.com/user-attachments/assets/9a2cade1-4462-4bed-b16c-aaa1b20bca02)

**Comment by maho3** [comment]
New 3 Gpc/h lightcone inference, using P(k) monopole with kmax=0.3
![image](https://github.com/user-attachments/assets/b0e7fc5d-5934-4bef-a6a5-30ab268bf877)
![image](https://github.com/user-attachments/assets/3acb0aec-ffba-4c11-8daf-7ac58ac03d3c)
![image](https://github.com/user-attachments/assets/2a4fbf36-581f-41f7-9877-259050b5a65f)
![image](https://github.com/user-attachments/assets/401e1c21-db5f-4bd7-844c-b806ce27b55c)
The Omega_m are suspiciously tight and the coverage is very good. This is possibly a bug, maybe the n(z) sampling is leaving a non-physical trace of the Omega_m. More investigation is needed.

**Comment by amariad** [comment]
<img width="419" alt="Image" src="https://github.com/user-attachments/assets/632189e3-0b73-4fd9-a0e7-1fee30e8ab9d" />

single posterior MTNG PkQk trained on z-dependent hod sims

**Comment by amariad** [comment]
<img width="664" alt="Image" src="https://github.com/user-attachments/assets/3cdb6f0f-ca36-4743-a028-5ae558291871" />

Same as above but only on Pk0

**Comment by maho3** [comment]
From Ana Maria's notes:

> ok, in path /anvil/scratch/x-adelgado1/gobig_fileshare you will find pytorch dictionaries named "MTNG_ensmble_model_dictionary_top10(name)_PkQk_{extendedHOD}_kmax(value).pt" I selected the top 10 models with narrowest 1-sigma range. In the files, “name” is either: “avg” for top 10 models with avg 1-sigma range of all cosmo params, “avgsig8" considering sigma_8 specifically, or “avgomsig8” considering the top 10 avg of Omega_m and sigma_8. I did these at kmax=0.2 and kmax=0.4.
> Ok I lied, 4 keys: 1.'Trained_posterior_ensemble_model' : the trained models, 2. 'Trained_model_summaries': the summaries output from the training, 3. 'Model_names_hyperparams': strings with each model and hyperparameters used in the ensemble and 4. 'Single_posterior_samples_pandas_dataframe': a pandas dataframe with the posterior samples directly used to make the plots.

I copied these to `/anvil/projects/x-phy240043/x-mho1/cmass-final/inference/mtng` so we have them on file.

And I was able to replicate the results! Here's using `MTNG_ensmble_model_dictionary_top10avgomsig8_PkQk_{extendedHOD}_kmax0.4.pt`

![Image](https://github.com/user-attachments/assets/80e93b32-0a29-4fba-ba4a-2b9404f4f4fb)

**Comment by maho3** [comment]
Here is the validation on the 3 Gpc/h Quijote Fiducial Simulation, using Halos. Training on CHARM Pk0 w/RSDs, Testing on Quijote 3 Gpc/h Pk0 w/RSDs.

Self-consistent predictions:
![Image](https://github.com/user-attachments/assets/d08f17cc-b483-405d-8fee-bba77e2924f8)

Prediction on 3 Gpc/h Quijote, Fiducial:
![Image](https://github.com/user-attachments/assets/2ef3493c-0b33-4504-a087-f685a764166c)

Prediction on our self-consistent simualtion, at the Fiducial Quijote cosmology:
![Image](https://github.com/user-attachments/assets/23e3ac0a-5b9d-4204-ba87-414000db674f)

**There seems to be some bias in our prediction with the full N-body 3 Gpc/h simulation**

![Image](https://github.com/user-attachments/assets/20044690-9e80-48bf-9b9f-b99bc5a7c5f6)

It seems that this is a result of power suppression at small-scales, and augmentation at large-scales.

![Image](https://github.com/user-attachments/assets/350fbe9b-9cd6-487a-89d8-3f13f88734ef)

I don't think this is an intrinsic CHARM problem, but it is because when we abundance-match halo masses (take the top # of halos, by mass), CHARM fails to produce the high-mass systems and as a result includes more low-mass systems (which produce more small-scale suppression in redshift-space)

Here's the differences in their minmasses after matching number density:
```
MTNGlike minmass: 13.408401
Quijote minmass: 13.42607186496521
```

**Comment by maho3** [comment]
This is another visualization of  the comparison between cosmology-matched CHARM and Quijote 3 Gpc/h. I am now subtracting off shot-noise and plotting it in semilogy

![Image](https://github.com/user-attachments/assets/f5a67a6b-0fe6-4781-9f5b-06ddc00a66b4)

And here's what happens when I match the mass ranges (so I am no longer fixing number density)

![Image](https://github.com/user-attachments/assets/6d77d401-5fbe-446c-b34b-1f3ccb68b29e)

So the mass cut is not the problem. I have the hypothesis that it is a noising problem, however. That applying noise to Quijote and to CHARM equivalently doesn't make the most sense

**Comment by maho3** [comment]
When validating MTNGlike on the real-space Pk0 for the Quijote 3 Gpc/h sim, a similar problem occurs:

![Image](https://github.com/user-attachments/assets/9bb768e0-68f2-4624-932a-0b16794b1b4a)

![Image](https://github.com/user-attachments/assets/9b5dd070-f16f-497b-8aa1-b6bd5d8b35a5)

![Image](https://github.com/user-attachments/assets/59c600f1-d5e1-4305-9480-fb669cf059e9)

Here are similar predictions for Pk024+Qk0:

![Image](https://github.com/user-attachments/assets/1aa374cf-b62a-4b6f-9d5d-4fec90dbc5b0)

![Image](https://github.com/user-attachments/assets/cc462066-0c7c-4d6d-beb1-b166d55a916f)

![Image](https://github.com/user-attachments/assets/1c0a8113-4f16-47f3-9cd2-d99eb6b6fda8)

**Comment by maho3** [comment]
Interestingly, when comparing these halo fields on 3 Gpc/h, there doesn't seem to be a ton of improvement when adding in higher-order clustering stats. If anything, the results get worse...

![Image](https://github.com/user-attachments/assets/095db9c1-3291-4a06-a5f8-dad26a3c4e3f)

![Image](https://github.com/user-attachments/assets/15e87f0f-6123-43f8-aec6-83feaec29e94)

This would suggest that the higher data dimensionality is struggling to recover the maximum constraining power. It's possible that compression or NLE methods would improve the constraints.

**Comment by maho3** [comment]
Now, when marginalizing over HODs, things look more consistent. Still training on CHARM and testing on Quijote 3 Gpc/h

#### zPk0
![Image](https://github.com/user-attachments/assets/58cf62cb-d4c2-4ba7-8d40-b63ced8b047f)

![Image](https://github.com/user-attachments/assets/c3247329-6dc0-44a5-8ab6-e255a7781a60)

#### nbar+zPk0

![Image](https://github.com/user-attachments/assets/f57ad18b-5eb7-4363-ad68-8b3bc2bc3925)

![Image](https://github.com/user-attachments/assets/6a40011f-80ee-4e5d-b714-a06b81ea64e5)

#### nbar+zPk024
![Image](https://github.com/user-attachments/assets/4aec451c-894d-4691-8a05-d0312a354a71)
![Image](https://github.com/user-attachments/assets/291f3780-c33d-4381-b038-d756a6d80bfd)

#### zPk024+zQk0
![Image](https://github.com/user-attachments/assets/2cf188d3-e412-4d05-b3e4-17ec602eb677)
![Image](https://github.com/user-attachments/assets/810016f3-f695-48bb-bc57-d7b3f461340e)

#### nbar+zPk024+zQk0
![Image](https://github.com/user-attachments/assets/056d4b96-160e-49da-b3c5-dda2c41867da)
![Image](https://github.com/user-attachments/assets/a0888199-85ea-49b0-b43f-98ec555e2fa3)

However, there's too little constraining power than we might expect in the Pk's.

**Comment by maho3** [comment]
Comparison of halo fields of Quijote 3 Gpc/h, CHARM 1 Gpc/h, and CHARM 3 Gpc/h

![Image](https://github.com/user-attachments/assets/d71fc403-41d4-4570-a304-652307289116)

**Comment by maho3** [comment]
### Optuna hyperparameter-tuned, train CHARM, test Quijote3Gpc/h
We've adapted our training procedure to now use Optuna. (See https://github.com/maho3/ltu-cmass/pull/115 )

We've also run more HODs per mtnglike simulation. There's now 5 HOD's per simulation. All results below are for galaxies, with RSDs, unless stated.

Similar to our results for abacuslike (https://github.com/maho3/ltu-gobig/issues/23#issuecomment-2973187672), we are unbiased for all metrics except the full bispectrum.

![Image](https://github.com/user-attachments/assets/d36def93-a406-469a-bbd5-a2409e398b97)

![Image](https://github.com/user-attachments/assets/11dbcd31-8618-4972-9678-10e38fe814b5)

![Image](https://github.com/user-attachments/assets/99dd4ae8-daf9-441f-95e4-12bc4bf2bcc8)

Here are the constraints for different HOD realizations, as a function of nbar

![Image](https://github.com/user-attachments/assets/d2c1d9e0-3306-4a2c-a103-cf10f9f5abac)

**Comment by maho3** [comment]
## Uniform+Gaussian Noising scheme, Jeffrey's prior, evaluating on Quijote 3 Gpc/h
Here, I'm training on CHARM simulations with the new noising scheme (Uniform+Gaussian) where the Gaussian's sigma_radial and sigma_transverse are set by a Jeffrey's prior where log(sigma) is uniform. Then, I'm testing on the single, real Quijote 3Gpc/h halos, where I don't add any noise (i.e. sigma_radial=sigma_transverse=0). All are done for ngc_lightcone summaries, with z-dependent HOD.

* Power spectra
![Image](https://github.com/user-attachments/assets/39083dfe-a079-4b15-a343-e5ee1545c069)
* Equilateral bispectrum
![Image](https://github.com/user-attachments/assets/58e266fe-d061-4daf-8883-97040f785130)
* Squeezed bispectrum
![Image](https://github.com/user-attachments/assets/74efbb7b-544a-46bb-9f11-fca1dccfe44c)
* Isosceles bispectrum
![Image](https://github.com/user-attachments/assets/149e01ab-8558-48ed-9ee8-9f9ec02cbaf0)
* Full bispectrum
![Image](https://github.com/user-attachments/assets/a62d2f96-6e1c-414e-b1bd-3758d23c6a87)

Takeaways:
* Omega_m, sigma8 seem unbiased in all cases! 
* We've seemed to stabilize the training. The degeneracies seem much sharper. This is likely a result of having 3k simulations, instead of 2k as in https://github.com/maho3/ltu-gobig/issues/23#issuecomment-3024711551 .
* We're still missing the degeneracy between h-n_s. Is there something that we're mismodeling?

**Comment by suicee** [comment]
## Exponential noise evaluating on Quijote 3 Gpc/h

* Pk0

![Image](https://github.com/user-attachments/assets/a10f0ca6-5358-46ff-85d6-43b9c043319d)

* Pk024

![Image](https://github.com/user-attachments/assets/5c42fb11-9285-4c76-b714-743cc9a1483f)

* Pk024+ Equilateral bispectrum

![Image](https://github.com/user-attachments/assets/9f95ad71-44f1-4c96-a56d-45761dcebadc)

**Comment by maho3** [comment]
## As before, with Gaussian+Uniform noise, but cutting on kmin=0.001
`kmin=0.001` is roughly the smallest k-mode you'd expect with a 1 Gpc/h volume

* Pk024
![Image](https://github.com/user-attachments/assets/49205c58-4bbe-452e-88de-6f8e106c7520)

* IsQk0
![Image](https://github.com/user-attachments/assets/1250e8b7-ee51-43c2-8865-8325d67ee620)
* Qk0
![Image](https://github.com/user-attachments/assets/c18b74a0-0ef0-4918-99e0-6b40a51cae6d)

n_s problem persists! Omega_m/sigma_8 looks good.

**Comment by maho3** [comment]
We fixed the n_s issue in https://github.com/maho3/ltu-gobig/issues/36.

Putting here some plots that I generated for the ILI meeting. At some point, we should come back and test this for all summaries.

## Uniform+Gaussian Noising scheme + NGP meshing, Jeffrey's prior, evaluating on Quijote 3Gpc/h w/ 4 Mpc/h noise
**Train on**: 3 Gpc/h CHARM+Zheng07ex with Uniform+Gaussian noise
**Noise prior**: Reciprocal(0.1, 4.5) Mpc/h
**Tracer**: ngc_lightcone
**Test on**: Quijote 3Gpc/h+Zheng07ex with 4 Mpc/h Gaussian noise

* Full Pk024+Qk0
<img width="540" alt="Image" src="https://github.com/user-attachments/assets/664f7a10-51d8-42e0-bfad-d402805e94c0" />
